tests/unit/fem/test_assemblediagonalpa.cpp

```diff
@@ -125,17 +125,19 @@ TEST_CASE("massdiag")
                    << std::pow(ne, dimension) << " elements." << std::endl;
          for (int order = 1; order < 5; ++order)
          {
-            Mesh * mesh;
+            Mesh mesh;
             if (dimension == 2)
             {
-               mesh = new Mesh(ne, ne, Element::QUADRILATERAL, 1, 1.0, 1.0);
+               mesh = Mesh::MakeCartesian2D(
+                         ne, ne, Element::QUADRILATERAL, 1, 1.0, 1.0);
```

Comment:
There is no copy-elision here, I think. It will better to use move assignment with `std::move`.

Reply:
That will already call the move assignment operator. I don't believe that `Mesh` has a copy assignment operator, so there is no danger of doing a copy by mistake.

Reply:
Unless it is deleted, the compiler defines the copy-assignment operator itself if it is not explicitly defined. Also, `object1 = object2` does not invoke the move-assignment operator -- for that you need to explicitly use `object1 = std::move(object2)`.

Reply:
The copy constructor is implicitly deleted because a move constructor is defined.

This is different from `object1 = object2` because instead of `object2`, we have the return value of a function, which is an rvalue not an lvalue.

Reply:
Even if the move constructor was not defined, `Mesh` would not have a copy assignment operator. For example, on the master branch, trying to copy-assign a mesh `mesh2 = mesh` gives:
```
ex1.cpp:112:10: error: object of type 'mfem::Mesh' cannot be assigned because its copy assignment operator is
      implicitly deleted
   mesh2 = mesh;
         ^
../mesh/../fem/../mesh/mesh.hpp:168:31: note: copy assignment operator of 'Mesh' is implicitly deleted because field
      'FaceElemTr' has a deleted copy assignment operator
   FaceElementTransformations FaceElemTr;
                              ^
../mesh/../fem/eltrans.hpp:508:20: note: copy assignment operator of 'FaceElementTransformations' is implicitly
      deleted because field 'FaceGeom' is of reference type 'Geometry::Type &'
   Geometry::Type &FaceGeom; ///< @deprecated Use GetGeometryType instead
```

Reply:
Judging by the error messages, the fact that the copy-assignment of `Mesh` is implicitly deleted is a coincidence -- due to the presence of a reference data member inside a data member.

In any case, we should not rely on the fact that the copy-constructor is deleted -- actually, it makes sense to implement the copy-assignment properly (e.g. using copy constructor + swap).

Regarding the assignment of (non-reference type) objects returned from a function -- you are right, these are rvalues (prvalues more precisely) and so `std::move` is not needed to invoke the move-assignment operator.

I also tested an expression of the form (this came up in another comment here: https://github.com/mfem/mfem/pull/2054#discussion_r578716262)
```c++
A a = cond ? fun(...) : fun(...);
```
(where `fun(...)` returns object of type `A` constructed in the return statement) and the compiler did use copy-elision and constructed `a` by invoking directly the constructor (no copy or move assignments or temporary objects).

This whole bussines with the lvalues, rvalues, xvalues, glvalues, and prvalues is tricky to learn and use correctly, so I consider this not-very-user-friendly for non-C++ experts (like myself).

Reply:
That the copy assignment operator is deleted in **this PR** is not a coincidence, it is because this PR defines a move constructor:
```
ex1.cpp:111:10: error: object of type 'mfem::Mesh' cannot be assigned because its copy assignment operator
      is implicitly deleted
   mesh2 = mesh;
         ^
../mesh/../fem/../mesh/mesh.hpp:503:4: note: copy assignment operator is implicitly deleted because 'Mesh'
      has a user-declared move constructor
   Mesh(Mesh &&mesh);
```

So even if the reference member data elsewhere was removed, the copy assignment operator would still be implicitly deleted — we can rely on that, it's not a coincidence. If we want to be extra sure (and more "self-documenting"), we can explicitly mark it as deleted.

The previous error messages were just to illustrate that even on the master branch, the copy assignment operator is deleted (for coincidental reasons, as you mentioned). What that tells us is that there is no user code out there using copy assignment for meshes. Also, I believe the default compiler-provided copy assignment would not do the right thing for some of the pointer data members.

As you mention, we could—if we wanted—implement a proper copy assignment operator for `Mesh`. I don't really think we should do this: meshes aren't the kind of thing you should be copying, and I don't believe there is user demand for it. Since meshes are expensive to copy, we would probably want to make the API very explicit, e.g. `mesh.CopyFrom(other_mesh)` or `mesh = Mesh::Clone(other_mesh)` rather than just `mesh = other_mesh`.

I completely agree with you that the "value category" semantics of C++ are **extremely** confusing, and I certainly don't have a good handle on them without looking it up in the reference manual. However, if the class has copy constructor marked explicit (as `Mesh` does) and deleted copy assignment operator (as `Mesh` does implicitly, we could make it explicit), then there is no risk of copying by mistake — the user doesn't need to know anything about lvalues, rvalues, or prvalues, it's just a guarantee of the class. If the user tries to do something that would cause a copy instead of a move, the compiler will provide an error message.

Reply:
This is another obscure C++ behavior -- adding move-assignment operator deletes the implicitly-defined copy-assignment. Since `Mesh` already has a copy-constructor, it is not a big stretch to add copy-assignment too but I don't see immediate need for that. However, if you don't want to add it, it will be better to explicitly delete the copy-assignment operator IMO.

Reply:
I explicitly deleted the copy assignment operator.

Reply:
One other thing I just noticed: I don't see the move-assignment operator in `ParMesh`. I think we should add it, similar to class `Mesh`. Maybe we can use a `= default` definition but that will require correct move-assignment operators for data members of types like `Array`, `Table`, etc.

The copy-assignment for `ParMesh` is implicitly deleted (e.g. because `ParMesh` has a user-defined move constructor, or because the copy-assignment is deleted in the base class `Mesh`). It will be better to make this explicit too.

Reply:
Ok, I created move assignment for ParMesh (so now move constructor and move assignment both use `ParMesh::Swap`), and I deleted the copy assignment operator.